### ISSUE TITLE
Questions  about the  PatchMixLoss

### ISSUE THREAD

**Issue by chumingqian** [raymin0223/patch-mix_contrastive_learning#5]
Hi, @raymin0223 ,
    
     May  i  have  ask you  2 questions about this  class;

```python
class PatchMixConLoss(nn.Module):
    def __init__(self, temperature=0.06):
        super().__init__()
        self.temperature = temperature

    def forward(self, projection1, projection2, labels_a, labels_b, lam, index, args):
        batch_size = projection1.shape[0]
        # proj: (bt, embed_dim)
        projection1, projection2 = F.normalize(projection1), F.normalize(projection2)

        # (bt, bt )
        anchor_dot_contrast = torch.div(torch.matmul(projection2, projection1.T), self.temperature)

        mask_a = torch.eye(batch_size).cuda() #  initial the  diag  matrix；
        mask_b = torch.zeros(batch_size, batch_size).cuda()  # zero mask；
        mask_b[torch.arange(batch_size).unsqueeze(1), index.view(-1, 1)] = 1 


        mask = lam * mask_a + (1 - lam) * mask_b


        logits_max, _ = torch.max(anchor_dot_contrast, dim=1, keepdim=True) #(bt, 1)
        logits = anchor_dot_contrast - logits_max.detach() # for numerical stability  (bt, bt)


        exp_logits = torch.exp(logits)  #(bt, bt)
        if args.negative_pair == 'diff_label':
            labels_a = labels_a.contiguous().view(-1, 1)
            logits_mask = torch.ne(labels_a, labels_a.T).cuda() + (mask_a.bool() + mask_b.bool())
            exp_logits *= logits_mask.float()

     
        log_prob = logits - torch.log(exp_logits.sum(1, keepdim=True))
        mean_log_prob_pos = (mask * log_prob).sum(1) / mask.sum(1)

   
        loss = -mean_log_prob_pos
        loss = loss.view(1, batch_size)

        loss = loss.mean()   
        return loss
```

   1.  In  this  class,  the  labels_b  has  not  been used.  And  in  the  total   project,   the ` args.negative_pair == 'all ' ` default  use  `all`  instead  the  `differ label`;    why  the  `labels_b `    and the  `  args.negative_pair == 'differ label'   `  has not  been  used;

  2.  As I understand it, the `i` th row of the variable `mask` matrix represents the `i `th sample of the generated patch, with the diagonal representing the percentage of patches that are kept from the original sample, and the other positions representing the percentage of patches that are used from another sample;   But the code after the variable mask, I really don't understand, can you explain the theory or recommend some related articles so that I can understand the following code;

Thank you very much for your work!

**Comment by raymin0223** [comment]
Hi @chumingqian,

Thanks for your interest in our work!

Here is the answer:

1. `diff_label` is to mimic [Supervised Contrastive Learning](https://arxiv.org/pdf/2004.11362.pdf); I assumed that it will improve performance if we make representations farther only for samples with different labels (_i.e._, exclude same label samples from negative pairs).         
Even though there will be lots of methods to define "same" or "different" labels due to the PatchMix, we only tested with `labels_a`, the original labels. So, it is not strict and correct method of `diff_labels`. 

2. Yes, `mask` exactly represents what you said. Because we mix samples with `lam` ratio, we assign it when making mask with `mask_a` and `mask_b`.       
We can say that we are using the mask as the label of CrossEntropy Loss.
As you know, in normal CE loss, we are using one-hot vector as ground truth; so, the equation looks like **$1*log(p)$**.
With the PatchMix, we replace that ground truth to **$\lambda*log(p_a) + (1- \lambda )*log(p_b)$**, cause the mixed sample is made by the interpolation between two samples.      
Then, `logits_max` is just for scaling logits, because previous work reported numerical stability issue.
And, `anchor_dot_contrast` is same with $(h(\tilde{z}_i)^{\top} z_i / \tau)$, `log_prob` is the calculation of Eq.2.
`mask * log_prob` is same with multiplying $\lambda$ or $(1- \lambda)$, and `mask.sum(1)` denote $\frac{1}{|I|}$.

You can refer to the above paper code and [MixCo](https://github.com/Lee-Gihun/MixCo-Mixup-Contrast) code.

Thanks and please leave any other question here!

**Comment by chumingqian** [comment]
Hi,  @raymin0223 ,

 Thank you very much for your prompt and detailed response. And it helped me.

My understanding is that when you calculate contrastive loss, it is done at the feature level, rather than by calculating the loss using the labels of mixed samples.

Let's assume you have two samples, $x_i$ and $x_k$, which belong to different classes, and the features learned by the network for these samples are $v_i$ and $v_k$,   respectively. Here, $\lambda$ and $1- \lambda$ represent the ratio of patches belonging to $x_i$ and $x_k$ respectively.

By using mix-up to combine their patches, you obtain a mixed sample, $x_{mix}$, and the features obtained from the network for this mixed sample are denoted as $v_{mix}$, 

According to your explanation, the features $v_{mix}$ should have a similarity ratio of $\lambda$ with $v_i$ and a similarity ratio of $1- \lambda$ with $v_k$.

If this is the case, I am curious about the following scenario: 
Suppose one sample is "normal," and the other sample is "crackle." 
In reality, even in the crackle audio, most of it consists of normal components due to the short duration of crackle, which is only about 25ms. 
If, in this case, (1- $\lambda$) of the patches are from the normal part, theoretically, $v_i$, $v_k$, and $v_{mix}$ should all represent the same type of feature, representing normal features. 

What would be the effect of performing contrastive learning in this situation?

**Comment by raymin0223** [comment]
Hi @chumingqian,

Thank your for your nice question.

In that case, our model will have better classification performance if it can output $v_k$ that includes the crackle information.
To optimize our loss function, the model may aim to output $v_i$ and $v_k$ totally different, although crackle samples also contains normal components.
So, we hope the model to be able to capture the short duration of crackle by contrasting normal and crackle samples.

**Comment by chumingqian** [comment]
Hi, @raymin0223 
Appreciate   a lot   and  i  am  trying   to  use  your   method,！:)

**Comment by chumingqian** [comment]
Hi,   @raymin0223 ,

   I'm back   :) ,  I have two questions (again ) for you regarding the code you demonstrated in the contrastive learning process.

  ```python
            elif args.method == 'patchmix_cl':
                features = model(images)
                output = classifier(features)
                loss = criterion[0](output, labels) 

                if args.target_type == 'grad_block':
                    proj1 = deepcopy(features.detach())
                elif args.target_type == 'grad_flow':
                    proj1 = features
                elif args.target_type == 'project_block':
                    proj1 = deepcopy(projector(features).detach())
                elif args.target_type == 'project_flow':
                    proj1 = projector(features)

                # use 'patchmix_cl' for augmentation
                mix_images, labels_a, labels_b, lam, index = model(images, y=labels, patch_mix=True, time_domain=args.time_domain)

                proj2 = projector(mix_images)
                loss += args.alpha * criterion[1](proj1, proj2, labels, labels_b, lam, index, args)

  ```


1. In your experiment, only the mixed features `mix_images` were passed through the projector, while `features` were obtained through `proj1= deepcopy(features.detach())` and did not pass through the projector. Why didn't you let both `features` and `mix_images` go through the `projector`?

2. Additionally, I am curious about the purpose of the experiments with the other three `target_type`. What were you trying to verify with them, and how effective were they?

**Comment by raymin0223** [comment]
Hi @chumingqian,

Sorry for the late reply, I'm now attending the conference.

1. When you see Table 2 results, we found that not using projector for target features showed the best results on ICBHI dataset. But, it can be different on other datasets.

2. Using projector or not has been discussed in traditional self-supervised learning research. I believe that projector performs like attention because model will learn the most important dimension to contrast samples. Thus, I experimented some target_types to confirm their effects.

**Comment by chumingqian** [comment]
Ah, I finally understand the content expressed in Table 2. Moreover, your perspective is correct. I am currently validating two different methods on the ICBHI dataset:

1. On Method 1, using 'grad_block' shows the best results.
2. Whereas on Method 2, employing 'project_flow' demonstrates the best performance.

This might indicate that 'projector' yields different outcomes for distinct algorithms.

 Finally,  wish you have  an enjoyable participation at the conference.
